source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -293,6 +311,43 @@ void SslSocket::onConnected() { ASSERT(info_->state() == Ssl::SocketState::PreHa
 
 Ssl::ConnectionInfoConstSharedPtr SslSocket::ssl() const { return info_; }
 
+void SslSocket::keylogCallback(const SSL* ssl, const char* line) {
+  ASSERT(ssl != nullptr);
+  auto config = static_cast<Ssl::ContextConfig*>(SSL_get_ex_data(ssl, ssl_ex_data_config_index_));
+  auto callback = static_cast<Network::TransportSocketCallbacks*>(
+      SSL_get_ex_data(ssl, ssl_ex_data_callback_index_));
+  auto access_log =
+      static_cast<AccessLog::AccessLogFile*>(SSL_get_ex_data(ssl, ssl_ex_data_file_index_));
+  ASSERT(config != nullptr);
+  ASSERT(callback != nullptr);
+  ASSERT(access_log != nullptr);
+  auto match =
+      tlsKeyLogMatch(callback->connection().connectionInfoProvider().localAddress(),
+                     callback->connection().connectionInfoProvider().remoteAddress(), *config);
+  if (match) {
+    access_log->write(line);
+  }
+}
+
+void SslSocket::disableTlsKeyLog(void) {
+  SSL_CTX* ctx = SSL_get_SSL_CTX(rawSsl());
```

Comment:
As commented previously, all of this code needs to move to the context_impl. https://github.com/envoyproxy/envoy/pull/19182#discussion_r802895651

Reply:
@ggreenway ,  context_impl does not have the connection information so that keylogCallback can not filter the local/remote.  enableTlsKeyLog should add the connection info as the SSL ex data.

Reply:
You cannot make changes to the SSL_CTX while handling a connection like this. This must move to the context.

The context is shared between many connections and many threads. It does not work at all to enable/disable this per-connection here. It will be flaky and/or crash.

You will register the keylog callback if keylogging is configured. In the callback for each connection, you'll need to apply the filtering. That may involve a call into the SslSocket if needed. 

There is already enough state stored on the SSL_CTX and SSL to get these two objects, so you will not need to register any more indexes with BoringSSL.

Reply:
@ggreenway, you said "There is already enough state stored on the SSL_CTX and SSL to get these two objects".
Do you mean we can get connection info from SSL ?  But I cannot find the direct way to get connection info from SSL. I can only find the API SSL_get_fd to get the fd. Could you help give more hints?

I can move enable/disable into context_impl, but I still need to use SSL_set_ex_data to register the  access logger info so that keylogcallback function can get the access logger from SSL.  If we can not get connection from SSL, I have to use SSL_set_ex_data to register.

Reply:
Sorry, I was incorrect about there being state stored in an index already; my recollection doesn't match reality.

I think you'll want to register one new index in ssl_socket, and store the `SslSocket*` in it in the constructor of `SslSocket`. In the callback, you can then retrieve the `SslSocket`, and the it has a pointer back to the `ctx_`. Then you can access the logger on the context. Does that give you access to everything you need?

Reply:
Firstly I tried to register an index to store SslSocket, but there is a loop dependency issue.  keylogCallback function is a static member function of class ContextImpl in context_impl.cc.  In function keylogCallback, it will use static_cast<SslSocket *>SSL_get_ex_data() to restore the SslSocket pointer. So context_impl.cc must include the header file "ssl_socket.h".  From source/extensions/transport_sockets/tls/BUILD we can see that ssl_socket_lib will depend on context_lib, so I can not add the header file ssl_socket.h into context_impl.cc.  I do not have good solution for this issue.  Instead, I register two indexes  for TransportSocketCallbacks and ContextConfig